src/kubernetes_charm.py

```diff
@@ -51,11 +61,18 @@
 class KubernetesRouterCharm(abstract_charm.MySQLRouterCharm):
     """MySQL Router Kubernetes charm"""
 
+    _PEER_RELATION_NAME = "mysql-router-peers"
+    _SERVICE_PATCH_TIMEOUT = 5 * 60
```

Comment:
we want to ensure that endpoints are actually connectable before we return them to the client (at the end of the event, we update data_interfaces and trigger a relation_changed on the client). this is similar to ensuring the mysqlrouter is running before event handler exits when enabling mysqlrouter

on cloud certain clouds, it can take a while for the service to be correctly created (along with the loadbalancer). will change so `_wait_until_service_reconciled` is only called on config-changed event so we are not potentially checking for every event

Reply:
Sleeping up to 5 minutes might produce more issues.
Do we have other options here? Time for Pebble notices?

Reply:
can we use defer here?

Reply:
i am leaning towards carl's second option of checking if the service is created before updating endpoints
however, this means that in the worst case, we may wait `update-status-hook-interval` until we update the endpoints in the databag (even though the endpoint is ready much earlier)

Reply:
refactored in ab4977647b51d285006bb26c8638090fb7926311. now, we only update_endpoints if they are connectable. otherwise we will do so in a future event when they become connectable

Reply:
On the server we are erroring out the hook after 60 seconds (generally ok for a `ClusterIP` service) with a timeout error.
So hook retry will eventually resolve it (or a `juju resolve`), if it's just a timing issue.

So, if endpoints are not connectable we will keep erroring out indefinitely. IMO it's the correct approach, even though ugly, since it renders the server useless if a client cannot use it, while allowing the manual retries (with `juju resolve`)

Reply:
i actually did refactor the code so that it no longer blocks for 5mins. instead, when we make a change to the service, we will update the endpoints in the relation databag when the next hook that runs (or this hook) where the service is connectable. i will remove the unused variable shortly

Reply:
on refresh, does juju retry the hook after, before or lost the errored event?

I'm ok with retrying in next hook

Reply:
actually, what error are you guys referring to?

if related to service connectivity check, the check itself does not raise any errors. instead, it checks if the service socket is connectable, and if so, sets endpoints in the databag. else, no-ops and will check during the next event that calls `reconcile`

Reply:
@paulomach @carlcsaposs-canonical i have some additional info on this topic

`lightkube_client.apply` (or any other lightkube client method) is a blocking method (does not return until the change has been made). as far as i can see, we cannot make this call non-blocking. this means that config-changed will run for a while until the service has been created and returns to the lightkube client

however, i also tested deleting a pod while the lightkube client was waiting for the service to be created. with the code as it is right now, the charm recovered. the service continued to be created once the lightkube client disconnected and pod restarted. upon restart, the endpoints get updated in the databag correctly when the next hook runs. i think that it is not worth overthinking here for refresh cases, as we can recover from them correctly

Reply:
you're right, the api request blocks until the resource has been created

im trying to explain why there is no issue at all - the api request blocks and with the current implementation, there is no downside to the pod getting refreshed/deleted/rescheduled. the reconcile approach ensures that the databag is correctly updated with the new endpoints during the next event

i think it would be irrelevant if we used apply (with server-side apply) or create, since we would only be applying/creating resources from the leader unit of the router app